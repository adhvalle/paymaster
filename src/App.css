.app__header {
  padding: var(--size-base-3x) 0;
  border-bottom: 1px solid var(--color-primary);
  background-color: var(--color-bg-primary);
}

.app__title {
  font-size: var(--font-size-2xl);
  color: var(--color-txt-primary);
  text-align: center;
}

.app__subtitle {
  font-size: var(--font-size-l);
  color: var(--color-txt-highlight);
}

.app__wrapper {
  padding: var(--size-base-6x) 0;
  border-bottom: 1px solid #6cb4b8;
  background-color: var(--color-bg-secondary);
}

.app__inner {
  display: grid;
  gap: var(--size-base-6x);
  grid-template-columns: repeat(2, 1fr);
  max-width: var(--max-width-md);
  margin: 0 auto;
}

.app__inner-header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: var(--size-base-4x);
}

/* MODAL */
.modal {
  position: absolute;
  width: 100vw;
  height: 100vh;
  top: 0;
  left: 0;
  display: grid;
  place-items: center;
  background-color: rgba(0, 0, 0, 0.7);
}

.modal__inner {
  width: 100%;
  max-width: var(--max-width-xs);
  padding: var(--size-base-3x) var(--size-base-6x);
  border: 2px solid var(--color-primary);
  border-radius: 10px;
  background-color: var(--color-bg-primary);
}

/* FORM */
.form {
  display: flex;
  flex-direction: column;
  gap: var(--size-base-2x);
  max-width: 320px;
  margin: 0 auto;
}

.form__label {
  display: block;
  font-size: var(--font-size-m);
  font-weight: semi-bold;
  margin-bottom: var(--size-base)
}

.form__input {
  width: 100%;
}
