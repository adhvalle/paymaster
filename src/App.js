import { useState } from 'react'
import './App.css'
import payments from './mock.json'
import { getSessionStorage, setSessionStorage } from './helpers/storage'
import Payments from './components/Payments'
import Balance from './components/Balance'
import ModalFriend from './components/ModalFriend'
import ModalExpense from './components/ModalExpense'

const PAYMENTS_KEY = 'payments'

function App () {
  const [addFriend, setAddFriend] = useState(false)
  const [addExpense, setAddExpense] = useState(false)
  const [paymentsList, setPaymentsList] = useState(() => {
    const storagedPayments = getSessionStorage(PAYMENTS_KEY)
    if (storagedPayments) return JSON.parse(storagedPayments)
    return payments.results
  })

  const _setNewData = data => {
    setPaymentsList(data)
    setSessionStorage(PAYMENTS_KEY, JSON.stringify(data))
  }

  const handleAddFriend = newName => {
    const newList = [...paymentsList]
    newList.push({ name: newName, payments: [] })
    _setNewData(newList)
    setAddFriend(false)
  }

  const handleAddExpense = newExpense => {
    const newList = [...paymentsList]
    const selectedFriend = paymentsList.findIndex(({ name }) => name === newExpense.name)
    newList[selectedFriend].payments.push({
      amount: newExpense.amount,
      created: newExpense.created,
      description: newExpense.description
    })
    _setNewData(newList)
    setAddExpense(false)
  }

  return (
    <main
      className='app'
      data-testid='app'
    >
      <header className='app__header'>
        <h1 className='app__title'>PAGADOR</h1>
      </header>
      <section className='app__wrapper'>
        <div className='app__inner'>
          <section>
            <div className='app__inner-header'>
              <h2 className='app__subtitle'>Balance</h2>
              <button onClick={() => setAddFriend(true)}>Añadir amigo</button>
            </div>
            <Balance payments={paymentsList} />
          </section>
          <section>
            <div className='app__inner-header'>
              <h2 className='app__subtitle'>Gastos</h2>
              <button onClick={() => setAddExpense(true)}>Añadir gasto</button>
            </div>
            <Payments payments={paymentsList} />
          </section>
        </div>
      </section>

      <ModalFriend
        showModal={addFriend}
        closeModal={() => setAddFriend(false)}
        addFriend={handleAddFriend}
      />

      <ModalExpense
        showModal={addExpense}
        payments={paymentsList}
        closeModal={() => setAddExpense(false)}
        addExpense={handleAddExpense}
      />
    </main>
  )
}

export default App
