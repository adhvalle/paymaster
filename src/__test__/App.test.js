import { render, screen, fireEvent, waitFor } from '@testing-library/react'
import App from '../App'

beforeEach(() => {
  render(<App />)
})

test('It should render the component', () => {
  const app = screen.getByTestId('app')
  expect(app).toMatchSnapshot()
})

test('It should open ModalFriend', async () => {
  fireEvent.click(screen.getByText('Añadir amigo'))
  await waitFor(() => expect(screen.getByText('Nombre')).toBeInTheDocument())
})

test('It should open ModalExpense', async () => {
  fireEvent.click(screen.getByText('Añadir gasto'))
  await waitFor(() => expect(screen.getByText('Amigo')).toBeInTheDocument())
})
