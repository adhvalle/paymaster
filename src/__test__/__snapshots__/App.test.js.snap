// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should render the component 1`] = `
<main
  class="app"
  data-testid="app"
>
  <header
    class="app__header"
  >
    <h1
      class="app__title"
    >
      PAGADOR
    </h1>
  </header>
  <section
    class="app__wrapper"
  >
    <div
      class="app__inner"
    >
      <section>
        <div
          class="app__inner-header"
        >
          <h2
            class="app__subtitle"
          >
            Balance
          </h2>
          <button>
            Añadir amigo
          </button>
        </div>
        <div
          class="balance"
        >
          <div
            class="balance__item"
          >
            <span>
              Chicho
            </span>
            <span
              class="balance__amount"
            >
              46,02 €
            </span>
          </div>
          <div
            class="balance__item"
          >
            <span>
              Nano
            </span>
            <span
              class="balance__amount balance__amount--negative"
            >
              -18,56 €
            </span>
          </div>
          <div
            class="balance__item"
          >
            <span>
              Cani
            </span>
            <span
              class="balance__amount balance__amount--negative"
            >
              -27,46 €
            </span>
          </div>
        </div>
      </section>
      <section>
        <div
          class="app__inner-header"
        >
          <h2
            class="app__subtitle"
          >
            Gastos
          </h2>
          <button>
            Añadir gasto
          </button>
        </div>
        <div
          class="payments"
        >
          <div
            class="payments__item"
          >
            <div
              class="payments__info"
            >
              <p>
                Cani
              </p>
              <p
                class="payments__item--rtl"
              >
                15,60 €
              </p>
            </div>
            <div
              class="payments__data"
            >
              <p>
                Palomitas
              </p>
              <p
                class="payments__item--rtl"
              >
                hace 9 días
              </p>
            </div>
          </div>
          <div
            class="payments__item"
          >
            <div
              class="payments__info"
            >
              <p>
                Nano
              </p>
              <p
                class="payments__item--rtl"
              >
                24,50 €
              </p>
            </div>
            <div
              class="payments__data"
            >
              <p>
                Cine
              </p>
              <p
                class="payments__item--rtl"
              >
                hace 12 días
              </p>
            </div>
          </div>
          <div
            class="payments__item"
          >
            <div
              class="payments__info"
            >
              <p>
                Chicho
              </p>
              <p
                class="payments__item--rtl"
              >
                29,83 €
              </p>
            </div>
            <div
              class="payments__data"
            >
              <p>
                Bolos
              </p>
              <p
                class="payments__item--rtl"
              >
                hace 13 días
              </p>
            </div>
          </div>
          <div
            class="payments__item"
          >
            <div
              class="payments__info"
            >
              <p>
                Chicho
              </p>
              <p
                class="payments__item--rtl"
              >
                59,25 €
              </p>
            </div>
            <div
              class="payments__data"
            >
              <p>
                Cena
              </p>
              <p
                class="payments__item--rtl"
              >
                hace 13 días
              </p>
            </div>
          </div>
        </div>
      </section>
    </div>
  </section>
</main>
`;
